File: backend/storage-service/src/Fix.Storage.Presentation/Mappers/FileContractMapper.cs
using Fix.Storage.Application.Files.Dtos;
using Google.Protobuf;
using Contract = Fix.Contracts.V1;

namespace Fix.Storage.Presentation.Mappers;

internal static class FileContractMapper
{
    public static Contract.File ToContract(this FileDto file)
    {
        var contract = new Contract.File
        {
            Id = file.Id.ToString(),
            Kind = file.Kind.ToContract<Contract.FileKind>(),
            Status = file.Status.ToContract<Contract.FileStatus>(),
            FileName = file.FileName,
            ContentType = file.ContentType,
            SizeBytes = file.SizeBytes,
            OrganizationId = file.OrganizationId.ToString(),
            UploadedBy = file.UploadedBy.ToString(),
            UploadedAt = file.UploadedAt.ToContract(),
            TotalLines = file.TotalLines,
            ProcessedLines = file.ProcessedLines,
            SucceededLines = file.SucceededLines,
            FailedLines = file.FailedLines,
            StorageUrl = file.StorageUrl,
        };

        if (file.FinishedAt is { } finished) contract.FinishedAt = finished.ToContract();
        if (file.Error is { } error) contract.Error = error;
        return contract;
    }

    public static Contract.FileLine ToContract(this FileLineDto line)
    {
        var contract = new Contract.FileLine
        {
            Id = line.Id.ToString(),
            Number = line.Number,
            Status = line.Status.ToContract<Contract.FileLineStatus>(),
            Values = { line.Values.ToDictionary() },
        };

        if (line.Message is { } message) contract.Message = message;
        if (line.ResultCode is { } code) contract.ResultCode = code;
        if (line.ProcessedAt is { } processed) contract.ProcessedAt = processed.ToContract();
        return contract;
    }

    public static Contract.FileKindSummary ToContract(this FileKindSummaryDto summary)
    {
        var contract = new Contract.FileKindSummary
        {
            Kind = summary.Kind.ToContract<Contract.FileKind>(),
            Total = summary.Total,
            Processing = summary.Processing,
            WithErrors = summary.WithErrors,
            CanUpload = summary.CanUpload,
        };

        if (summary.LastUploadedAt is { } last) contract.LastUploadedAt = last.ToContract();
        return contract;
    }

    public static Contract.FileTemplate ToContract(this FileTemplateDto template) => new()
    {
        Kind = template.Kind.ToContract<Contract.FileKind>(),
        Processed = template.Processed,
        AcceptedExtensions = { template.AcceptedExtensions },
        Columns =
        {
            template.Columns.Select(c => new Contract.FileTemplateColumn
            {
                Name = c.Name,
                Required = c.Required,
                Description = c.Description,
                Example = c.Example,
            }),
        },
        ExampleCsv = ByteString.CopyFrom(template.ExampleCsv),
    };

    public static Contract.FileDownloadUrl ToContract(this FileDownloadUrlDto download) => new()
    {
        Url = download.Url,
        ExpiresAt = download.ExpiresAt.ToContract(),
    };
}

File: frontend/bff/src/infrastructure/grpc/mappers/file.contract-mapper.ts
import type { FileDownloadUrlDto } from '../../../application/files/dtos/file-download-url.dto.js';
import type { FileKindSummaryDto } from '../../../application/files/dtos/file-kind-summary.dto.js';
import type { FileLineDto } from '../../../application/files/dtos/file-line.dto.js';
import type { FileTemplateDto } from '../../../application/files/dtos/file-template.dto.js';
import type { FileDto } from '../../../application/files/dtos/file.dto.js';
import { nullable } from './common.contract-mapper.js';
import { fileKindEnum, fileLineStatusEnum, fileStatusEnum } from './enum.contract-mapper.js';

export interface ContractFile {
  id: string;
  kind: string;
  status: string;
  fileName: string;
  contentType: string;
  sizeBytes: string;
  organizationId: string;
  uploadedBy: string;
  uploadedAt: string;
  finishedAt?: string;
  totalLines: number;
  processedLines: number;
  succeededLines: number;
  failedLines: number;
  error?: string;
  storageUrl: string;
}

export interface ContractFileLine {
  id: string;
  number: number;
  status: string;
  values: Record<string, string>;
  message?: string;
  resultCode?: string;
  processedAt?: string;
}

export interface ContractFileKindSummary {
  kind: string;
  total: number;
  processing: number;
  withErrors: number;
  lastUploadedAt?: string;
  canUpload: boolean;
}

export interface ContractFileTemplate {
  kind: string;
  processed: boolean;
  acceptedExtensions: string[];
  columns: { name: string; required: boolean; description: string; example: string }[];
  exampleCsv: Buffer;
}

export const toFileDto = (f: ContractFile): FileDto => ({
  id: f.id,
  kind: fileKindEnum.fromContractRequired(f.kind),
  status: fileStatusEnum.fromContractRequired(f.status),
  fileName: f.fileName,
  contentType: f.contentType,
  sizeBytes: Number(f.sizeBytes),
  organizationId: f.organizationId,
  uploadedBy: f.uploadedBy,
  uploadedAt: f.uploadedAt,
  finishedAt: nullable(f.finishedAt),
  totalLines: f.totalLines,
  processedLines: f.processedLines,
  succeededLines: f.succeededLines,
  failedLines: f.failedLines,
  error: nullable(f.error),
  storageUrl: f.storageUrl,
});

export const toFileLineDto = (l: ContractFileLine): FileLineDto => ({
  id: l.id,
  number: l.number,
  status: fileLineStatusEnum.fromContractRequired(l.status),
  values: { ...l.values },
  message: nullable(l.message),
  resultCode: nullable(l.resultCode),
  processedAt: nullable(l.processedAt),
});

export const toFileKindSummaryDto = (k: ContractFileKindSummary): FileKindSummaryDto => ({
  kind: fileKindEnum.fromContractRequired(k.kind),
  total: k.total,
  processing: k.processing,
  withErrors: k.withErrors,
  lastUploadedAt: nullable(k.lastUploadedAt),
  canUpload: k.canUpload,
});

export const toFileTemplateDto = (t: ContractFileTemplate): FileTemplateDto => ({
  kind: fileKindEnum.fromContractRequired(t.kind),
  processed: t.processed,
  acceptedExtensions: t.acceptedExtensions,
  columns: t.columns.map((c) => ({ ...c })),
  exampleCsv: Buffer.from(t.exampleCsv ?? []),
});

export const toFileDownloadUrlDto = (d: { url: string; expiresAt: string }): FileDownloadUrlDto => ({
  url: d.url,
  expiresAt: d.expiresAt,
});
